### ISSUE TITLE
Image uploader doesn't work if the field name is in multi-dimension array form

### ISSUE THREAD

**Issue by malikt** [magento/magento2#15675]
I have to get images from admin end for "Customizable Options", I added a new column to custom option type value table and i was able to inject "File" field through rewrite of Ui Dataprovider. The default name for the upload field makes as product[options][0][values][0][image]. But when i try to upload the image, i get an exception. however, If i rename the upload field  as a string  like "image" Or name like "option[image]" The upload and preview both work. 

### Preconditions

1. Magento 2.2.0
2. PHP 7.0.23

### Steps to reproduce

1. Add a new column to catalog_product_option_type_value table with any name.
2. rewrite Magento\Catalog\Ui\DataProvider\Product\Form\Modifier\CustomOptions in etc/adminhtml/di.xml and add the Upload field.
3. Entries in etc/di.xml for base temp path, allowed extensions etc
4. Edit any product and add customizable option like radio, checkbox or dropdown, you will see the Browse button.
5. upload any image.

### Expected result

1. Image should upload and show a preview and on saving the product should be saved.

### Actual result

1. It returns with an exception of undefined index.
![image](https://user-images.githubusercontent.com/24920883/40868798-29d7567e-662a-11e8-97d7-91488c627a04.png)

The problem lies with the Upload button name. If its multidemension Array form, then the vendor\magento\framework\File\Uploader.php instead of parsing the array it just takes the nested arrays as params.

**Comment by malikt** [comment]
Hi,
I got a hacky way to fix this, you guys may find a better way.
Instead of rewriting the core file I just copied following file to my module and I directed the controller and models to use it instead of core, In this way core files will be doing what they were:
1.` \Magento\Framework\File\Uploader `
as there is already a preg code to find if file id is in an array format at line around 535
`preg_match("/^(.*?)\[(.*?)\]$/", $fileId, $file);`
its checking if file id has [ and ] and this particular line causes issues. 
in case of `product[options][0][values][0][image]` , it results as array of "product" i.e. `$file[0]` and its key as "options][0][values][0][image" i.e. `$file[1]` which is not correct. so it returns an exception due to code at line 545
> foreach ($fileAttributes as $attributeName => $attributeValue) {
>              $tmpVar[$attributeName] = $attributeValue[$file[1]];
>  }

so i used another preg to match if index contains any array brackets, and if yes then traverse the `$attributeValue` again and find the image name, i replaced the foreach as following:

```
foreach ($fileAttributes as $attributeName => $attributeValue) {
    preg_match("/^(.*?)\](.*?)\$/", $file[1], $wasMultiArray);
    if ( is_array($attributeValue) &&
        count($wasMultiArray) > 0 &&
        count($wasMultiArray[0]) > 0 &&
        count($wasMultiArray[1]) > 0 ) {
        $imageName = "";
        array_walk_recursive( $attributeValue,
            function ( $item, $key ) use ( &$imageName )
            {
                $imageName = $item;
            }
        );
        $tmpVar[$attributeName] = $imageName;
    }else{
        $tmpVar[$attributeName] = $attributeValue[$file[1]];
    }
}
```
In my case i have to find if the index starts with "]" you may use this preg accordingly.

**Comment by engcom-backlog-nickolas** [comment]
Hello @malikt, thank you for your report.
We've acknowledged the issue and added to our backlog.

**Comment by fme-tahir** [comment]
Hi, any fix for this issue on Magento 2.2?
